docs/rfcs/027-pageserver-disaster-recovery.md

```diff
@@ -0,0 +1,157 @@
+  to figure out if there was a corruption of data (we would compare with what
+  the pageserver stores for the "official" timeline).
+* Using the WAL to arrive at historic pages we can then back up to S3 so that
+  WAL itself can be discarded, or at least not used for future replays.
+  Again, this sounds a lot like what the pageserver is already doing, but the
+  point is to provide a fallback to the service provided by the pageserver.
+
+## Design
+
+### Design constraints
+
+The main design constraint is that the feature needs to be *simple* enough that
+the number of bugs are as low, and reliability as high as possible: the main
+goal of this endeavour is to achieve higher correctness than the pageserver.
+
+For the background process, we cannot afford a downtime of the timeline that is
+being cloned, as we don't want to restrict ourselves to offline tenants only.
+In the scenario where we want to recover from disasters or roll back to a
+historic lsn through support staff, downtimes are more affordable, and
+inevitable if the original had been subject to the corruption. Ideally, the
+two code paths would share code, so the solution would be designed for not
+requiring downtimes.
+
+### API endpoints
+
+The proposed design of the new pageserver API endpoint is:
+
+```
+/v1/tenant/:tenant_id/timeline/:timeline_id/recover?from_lsn=<from_lsn>&to_lsn=<lsn>&target=<target_timeline_id>
+```
+
+The `recover` endpoint answers to `POST` requests and issues a WAL recovery
+process. It takes the WAL of the specified timeline, and writes its results
+into a *different* timeline, the target timeline. This two timeline scheme
+ensures that we can stil serve requests and preserves immutability guarantees
+assumed by the [generation numbers RFC], as in corruption situations we might
+indeed end up with different content than the original timeline we are storing
+data from.
+
+Where the parameters are:
+
+* `:tenant_id`: the id of the tenant to recover
+* `:timeline_id`: the id of the timeline to recover
+* `from_lsn`: The lsn at the start of our recovery process.
+  Optional, if not specified the recovery process is fully WAL based.
+  If specified, the recovery process starts from data at the *main* timeline
+  at the specific lsn (maybe we still trust that lsn). In any case, the target
+  timeline must be less advanced than `from_lsn`.
+* `to_lsn`: the end of the lsn range we want to recover. As usual, it's an
+  exclusive limit. This allows playing back the WAL to a specific moment in the
+  past, if this is wanted/needed. E.g. there might be a data-corrupting
+  pageserver crash at lsn 1234 and we don't want that so we play back until
+  1233.
+* `target`: the timeline id we are writing into. The parameter is optional;
+  if no target timeline is provided, a new timeline is created.
+  One specifies either both `from_lsn` and `target`, or neither.
+  Writing into the same timeline is forbidden.
+
```

Comment:
It would be good to define:
- the state of the system is when this API call returns (e.g. does the timeline "exist" from the point of view of a "GET /tenant/foo/timeline/bar" once we've called this API, or only later when the recovery completes?)
-  how the caller can check for progress/completion.
- after recovery is complete, does the timeline automatically transition to being a "normal" timeline that is readable by endpoints?  Or is there some extra step needed later?

Reply:
> the state of the system is when this API call returns

I suppose we want the API call to immediately return the timeline ID and then give the caller some means to track the status of the process.

We definitely will not be able to serve get requests for "current" lsns, but for lsns at past moments we would probably be able to. I think there are analogies here with asking for future lsns where WAL replay has not caught up yet? I'd model the process, also alluded to [in my reply above](https://github.com/neondatabase/neon/pull/5248#discussion_r1321832429):

* Start is special, takes reproducible initdb snapshot (cc https://github.com/neondatabase/neon/issues/2592). API call for new timelines also probably returns after initdb ran? idk.
* timeline is created with "wal from other timeline until lsn no `target`" status, probably saved somewhere in the metadata?
* pageserver does its thing where it replays the WAL to make pages
* timeline reaches the `target` lsn: now it changes its status to turn off the "wal from other timeline until lsn `target`" flag, and listens to WAL input for its own timeline ID.

> does the timeline "exist" from the point of view of a "GET /tenant/foo/timeline/bar" once we've called this API, or only later when the recovery completes
>
> how the caller can check for progress/completion.

The console has some kind of "operations" concept, but I don't think we need such in the pageserver: we can just return whatever lsn the process is currently at, and the lsn we want to arrive at is known to the caller, so they can derive a nice percent value in terms of progress from that.

> after recovery is complete, does the timeline automatically transition to being a "normal" timeline that is readable by endpoints? Or is there some extra step needed later?

I think we can just serve all read calls up to the lsn we have replayed.

Reply:
> If our recovery code is a specialized instantiation of a Timeline object with a new ID, then a bug in the pageserver might occur just the same during recovery as it did to begin with. I think that's okay, we just need to be consistent in how we think/talk about this feature

Good point. The goal is not to reimplement a simpler clone of the entire stack, say without compaction, but it's to use a non-buggy pageserver to undo the mess created by a buggy pageserver. Or even if there was no bug, cosmic rays are a thing (so are CT [bit](https://news.ycombinator.com/item?id=27728287) [flips](https://news.ycombinator.com/item?id=35883910)).

I will reword that part to make it clearer :+1: .